api_callable.go

```diff
@@ -8,68 +8,59 @@ import (
 	"google.golang.org/grpc/codes"
 )
 
-// Represents a GRPC call stub.
-type APICall func(context.Context, interface{}) (interface{}, error)
+// A user defined call stub.
+type APICall func(context.Context) error
 
 // scaleDuration returns the product of a and mult.
 func scaleDuration(a time.Duration, mult float64) time.Duration {
 	ns := float64(a) * mult
 	return time.Duration(ns)
 }
 
-// stubWithRetry returns a wrapper for stub with an exponential backoff retry
-// mechanism based on the values provided in retrySettings.
-func stubWithRetry(stub APICall, retrySettings retrySettings) APICall {
-	return func(ctx context.Context, req interface{}) (resp interface{}, err error) {
-		backoffSettings := retrySettings.backoffSettings
-		// Forces ctx to expire after a deadline.
-		ctx, _ = context.WithTimeout(ctx, backoffSettings.totalTimeout)
-
-		delay := backoffSettings.delayTimeoutSettings.initial
-		timeout := backoffSettings.rpcTimeoutSettings.initial
-
-		for {
-			// If the deadline is exceeded...
-			if ctx.Err() != nil {
-				return nil, ctx.Err()
-			}
-			timeoutCtx, _ := context.WithTimeout(ctx, backoffSettings.rpcTimeoutSettings.max)
-			timeoutCtx, _ = context.WithTimeout(timeoutCtx, timeout)
-			resp, err = stub(timeoutCtx, req)
-			code := grpc.Code(err)
-			if code == codes.OK {
-				return resp, err
-			}
-			if !retrySettings.retryCodes[code] {
-				return nil, err
-			}
-			delayCtx, _ := context.WithTimeout(ctx, backoffSettings.delayTimeoutSettings.max)
-			delayCtx, _ = context.WithTimeout(delayCtx, delay)
-			<-delayCtx.Done()
-
-			delay = scaleDuration(delay, backoffSettings.delayTimeoutSettings.multiplier)
-			timeout = scaleDuration(timeout, backoffSettings.rpcTimeoutSettings.multiplier)
+// invokeWithRetry calls stub using an exponential backoff retry mechanism
+// based on the values provided in retrySettings.
+func invokeWithRetry(ctx context.Context, stub APICall, retrySettings retrySettings) (err error) {
+	backoffSettings := retrySettings.backoffSettings
+	// Forces ctx to expire after a deadline.
+	childCtx, _ := context.WithTimeout(ctx, backoffSettings.totalTimeout)
+	delay := backoffSettings.delayTimeoutSettings.initial
+	timeout := backoffSettings.rpcTimeoutSettings.initial
+	for {
+		// If the deadline is exceeded...
+		if childCtx.Err() != nil {
+			return childCtx.Err()
+		}
+		timeoutCtx, _ := context.WithTimeout(childCtx, backoffSettings.rpcTimeoutSettings.max)
+		timeoutCtx, _ = context.WithTimeout(timeoutCtx, timeout)
+		err = stub(timeoutCtx)
+		code := grpc.Code(err)
+		if code == codes.OK {
+			return
```

Comment:
Agreed, try to avoid naked returns. I also wouldn't name the return value, to avoid temptation/confusion.